apps/admin/backend/src/app.ts

```diff
@@ -548,92 +539,61 @@ function buildApi({
         input.writeInId
       );
       assert(castVoteRecordData);
-
-      const { contestId, optionId, electionId, cvr } = castVoteRecordData;
-
-      const electionRecord = store.getElection(electionId);
-      assert(electionRecord);
-
-      const { election } = electionRecord.electionDefinition;
-
-      const ballotStyle = getBallotStyle({
-        ballotStyleId: cvr._ballotStyleId,
-        election,
-      });
+      const { contestId, optionId, cvr } = castVoteRecordData;
       if (cvr._layouts === undefined || cvr._ballotImages === undefined) {
         return []; // The CVR does not have ballot images.
       }
-      if (!ballotStyle) {
-        throw new Error('unexpected types');
-      }
-      const allContestIdsForBallotStyle = getContests({
-        ballotStyle,
-        election,
-      }).map((c) => c.id);
-      const [layouts, ballotImages] = iter(cvr._layouts)
-        .zip(cvr._ballotImages)
-        .toArray()
-        .sort(([a], [b]) => a.metadata.pageNumber - b.metadata.pageNumber)
-        .reduce<[BallotPageLayout[], InlineBallotImage[]]>(
-          ([layoutsAcc, ballotImagesAcc], [layout, ballotImage]) => [
-            [...layoutsAcc, layout],
-            [...ballotImagesAcc, ballotImage],
-          ],
-          [[], []]
+
+      // Identify the page layout, contest layout, and associated image.
+      let pageLayout: Optional<BallotPageLayout>;
+      let contestLayout: Optional<BallotPageContestLayout>;
+      let ballotImage: InlineBallotImage | null | undefined;
+      for (const [index, layout] of cvr._layouts.entries()) {
+        if (layout === null) continue;
+
+        const foundContestLayout = layout.contests.find(
+          (contest) => contest.contestId === contestId
         );
```

Comment:
We're now using the changes in #3180 to find a layout without necessarily having all the layouts, and iterating through them.